extensions/amp-video/0.1/video-cache.js

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright 2021 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Services} from '../../../src/services';
+import {createElementWithAttributes, matches} from '../../../src/dom';
+import {extensionScriptsInNode} from '../../../src/service/extension-script';
+import {resolveRelativeUrl} from '../../../src/url';
+import {toArray} from '../../../src/core/types/array';
+import {user} from '../../../src/log';
+
+/**
+ * Add the caching sources to the video if opted in.
+ * @param {!Element} videoEl
+ * @param {!Window} win
+ * @return {!Promise}
+ */
+export function addCacheSources(videoEl, win) {
+  if (
+    !extensionScriptsInNode(win.document).some(
+      (extension) => extension.extensionId === 'amp-cache-url'
+    ) ||
+    !videoEl.querySelector('source[src]')
```

Comment:
What if it's a `<video src=""` with no source child? Please add this configuration in your unit tests as well, that'd be a pretty bad bug.

Reply:
Filed https://github.com/ampproject/amphtml/issues/33896 to support it